### DISCUSSION TITLE
confused set questions

### DISCUSSION THREAD

**Discussion by dizzwold** [awawa-dev/HyperHDR#890] [Q&A]
Hi Guys,

 I'm new to HyperHDR and relativity new to RPI's.
  I've found this great software and project from following Everything Smart Home's video guide;
https://www.youtube.com/watch?v=J26oYlKyq7Q&t=1133s
 I've got some SK6812 CW LEDs on the way, I have a RPi 4 and a few Zero's, Kinivo 840BN HDMI Splitter and the following Capture Card;
https://www.amazon.co.uk/dp/B08JHNKDP8?psc=1&ref=ppx_yo2ov_dt_b_product_details
 In the video guide above (Lewis (Everything Smart Home)), doesn't use an LED driver board and connect's the LED's directly to the RPi (albeit a Zero where I plan on using a RPi 4)
 Do I still need an LED driver?
 I would like the lowest latency and it's not a big deal for me to purchase a Pico if needed.
I'd be grateful for you feedback and help in gathering the correct parts for a grate experience.

**Comment by awawa-dev** [comment]
Hi
This method has not been officially supported in HyperHDR for a long time and you use it at your own risk, accepting: lack of the required level-shifter 3.3v to 5v for LEDs, problems with handling larger strips ~250-300 RGBW LEDs where usually the last segment may behave unstable, newer RPi such as RPi5 do not support such a solution anyway, it is necessary to run everything as root.
For beginners, I recommend HyperSerialPico and one of the Adafruit boards with a built-in level-shifter. HyperSPI has also been supporting them for some time now, the connection is simple as you can see in the screenshots on the project website, but it does not forgive any shortcuts or negligence: after all, we work with speeds ~20-25Mb. So you can start with HyperSerialPico with such a board and then switch to HyperSPI at any time if you find it necessary. Just remember that HyperSPI requires much shorter cables to connect Pico to RPi.

When it comes to delays, remember that TV also has a delay: the largest is usually in cinema mode, depending on the manufacturer, e.g.: 200ms (here, HyperHDR can be faster than TV, although it is usually not visible to the naked eye and it has to be recorded with a phone with a very high framerate) and smallest in game mode ~15ms. If you don't use very large LED strips, HyperSerialPico should be sufficient even for sport mode. Moreover, what matters more here is whether you use smoothing that itself introduces a slight delay.

I don't know the grabber you chose, loop grabbers are problematic, currently the standard is to use ms2130 and an Ezcoo splitter and this combination also supports eg. HDCP and Dolby Vision (in LLDV mode).

**Reply by dizzwold** [reply]
Hi Awawa-dev,

 Very grateful for your reply and thank you for taking the time to do so.
 I have electronics skills and many esp devices. I'm not a fan of most the esp boards as I've found them to be quite unreliable, I think this is mostly due to the construction on my building.
 Ironically I do have some SN74HCT245N kicking around, that where for HA led's project.
 I'm getting 144 leds/m SK6812, as my tv is around 200mm away from the wall.
 My capture card is stated to be a 4K pass through HDCP2.2 but untested as yet. But this could always be returned.
 My Splitter is 8K@60hz 4K@120hz HDR 10+, Dolby Vision, HDCP 2.3, 48Gbps
 So on paper would my splitter and capture card be sufficient?
 Smoothing?
 So will I need the RPi 4 and an Adafruit Pico running HyperSPI for the best results?
 As I'll have in the region of 450 leds, should I consider using a multi/parallel pico, if a pico is needed?

**Reply by awawa-dev** [reply]
> I'm getting 144 leds/m SK6812, as my tv is around 200mm away from the wall.

At such a long distance, I don't know if the colors will mix-up on the wall. LEDs emit the highest power in the range of ~+/-30 degrees

> My Splitter is 8K@60hz 4K@120hz HDR 10+, Dolby Vision, HDCP 2.3, 48Gbps

well, on paper it supports Dolby and your TV will probably receive the Dolby signal, but the question is what about the grabber. Grabbers can only handle LLDV which is very similar to HDR, Ezcoo splitters are rather unique in this respect and can force LLDV so that your TV and grabber will get LLDV and not some other DV protocol that your TV will support but the grabber will not (usually results in a pink image). Although this is a rather undocumented function of these splitters. You can find more here in the discussions.

> 450 RGBW leds

if you have a pico you can run the benchmark described on my blog and you'll get the results even without the LEDs connected. I think with a single-segment setup it will be around 60fps. So it's your decision if that's enough for you or to choose multi-segment setup.

**Reply by dizzwold** [reply]
By Benchmark, do you mean, https://github.com/awawa-dev/HyperHDR/releases/tag/v17.0.0.0 ?
I'll contact my splitter manufacture and enquire about the LLDV.

**Reply by awawa-dev** [reply]
no, this one https://www.hyperhdr.eu/2023/02/ultimate-guide-on-how-to-build-led.html#chapter9 but using Pico and HyperSerialPico

**Reply by awawa-dev** [reply]
> I'll contact my splitter manufacture and enquire about the LLDV.

For example, here is the DV failure of the Feintech splitter: https://github.com/awawa-dev/HyperHDR/discussions/729#discussioncomment-8126042

**Reply by dizzwold** [reply]
@Awawa-dev,

 Thank you, and thank you for your patience.
 I've emailed my splitter manufacturer and await there reply regarding the LLDV.
 The leds should be here in the next few days too
 I'm going to order a Adafruit Scorpio, then if I do need the extra outputs, I'll have them available.

**Reply by dizzwold** [reply]
@awawa-dev 

 I'll soon have the Pico (Adafruit Scorpio), and as I understand it I need to install HyperSerialPico on the Pico from here;
https://github.com/awawa-dev/HyperSerialPico

 As for installing HyperHDR on the Pi 4, I had seen another installation guide by Justin Nelson and that states to install RPi OS Lite 64, but then when I look at the Installation Page it states that "Deprecated since the latest Raspberry Pi OS based on Debian Bookworm"
 So how should I install the HyperHDR onto the Pi4?
I do have a spare TV, keyboard, mouse, etc if that helps, but I only have an IMac computer (no windows devices). 
Sorry if this sound dumb.

**Reply by awawa-dev** [reply]
> I'll soon have the Pico (Adafruit Scorpio), and as I understand it I need to install HyperSerialPico on the Pico from here;
https://github.com/awawa-dev/HyperSerialPico

that's right

> but then when I look at the Installation Page it states that "Deprecated since the latest Raspberry Pi OS based on Debian Bookworm"
So how should I install the HyperHDR onto the Pi4?

It's because Raspberry Pi org. became some time ago M$ itself this time by forcefully promoting their tools and gaining more telemetry data: https://github.com/awawa-dev/HyperHDR/issues/699 and here https://github.com/raspberrypi/bookworm-feedback/issues/72 (they have just locked the discussion... LOL) So there will be no more Bookworm-based HyperHDR images until something changes, only older Bullseye (that's why it's depreciated) with working wifi configuration. To use latest Bookworm images you need their imager tool. Fortunately for us, a much better alternative to RPi has appeared in the form of cheap minicomputers based on the N100 where we are not tied to one Linux distribution with their closed binary drivers.

**Reply by dizzwold** [reply]
@awawa-dev 

 So are you saying I can't use the Pi4 any-more?

**Reply by awawa-dev** [reply]
you can, it's about SD images

**Reply by dizzwold** [reply]
@awawa-dev 

 Okay, I've found this following page with a Raspberry Pi SD Image Installation guide;
https://www.hyperhdr.eu/2020/11/hyperhdr-prepare-for-building-buying_17.html

 So Should I follow these instructions using Balena Etcher and write the wpa_supplicant with BBEdit?

**Reply by awawa-dev** [reply]
Use official RPi imager guide to install latest Raspberry Pi OS on SD since they force users to use it. Remember to enable SSH. Then install on RPi HyperHDR using APT repository:  https://awawa-dev.github.io/

**Reply by dizzwold** [reply]
@awawa-dev 

 Thank you.

 I'll try to get it done tomorrow and see how I get on (novice). Hopefully the Pico will arrive tomorrow.

 Link somewhere "To Buy You A Coffee/Coffee's"?

**Reply by awawa-dev** [reply]
https://github.com/sponsors/awawa-dev 😉

**Reply by dizzwold** [reply]
@awawa-dev ,

 This might sound like a daft question, but can I use a desktop version RPi OS?

 I've never been able to ssh from my IMac to any Rpi, and I've spent ll day in trying again with no success.

 Don't ask why this is, but it's never worked, which is a real *isser.

**Reply by awawa-dev** [reply]
it's not recommended for rpi

**Reply by dizzwold** [reply]
@awawa-dev ,

 I've finally managed to ssh into the pi4.

 Any ideas on how to upload the HyperSerialPico to the Scorpio from an Imac?

 I've just tried with the Arduino IDE but it think it doesn't like the file format .uf2 of the  HyperSerialPico_sk6812Cold file.

 I can connect to the Scorpio via terminal on the IMac, but I don't have a clue how or if it can be flashed through Terminal.

 Just thought, to flash the Scorpio, I am connecting the Scorpio directly to my computer and not the Pi4 with HyperHDR on?

**Reply by awawa-dev** [reply]
https://github.com/awawa-dev/HyperSerialPico?tab=readme-ov-file#how-to-flash-it

**Reply by dizzwold** [reply]
@awawa-dev,

 Yeah, I've already got that page, but sorry to say I don't have any windows computers, so no file explorer.

**Reply by awawa-dev** [reply]
Doesn't matter

**Reply by dizzwold** [reply]
@awawa-dev 

 Okay, figured that part out;
Connect Scorpio to the IMac and press boot and reset, release reset and RPi will appear in sidebar of Finder under devices.
Drag and drop the folder required (HyperSerailPico_sk6812Cold in my case), to the RPi listed in the sidebar of Finder.

 Just gone back to Terminal after 5 minuets and typed "screen /dev/tty.usbmodemFA121", and I get the reply;

 Cannot open line  '/dev/tty.usbmodemFA121 for R/W:  Resource busy
 Sorry, could not find a PTY

 Does that mean it's completed or bricked?

**Reply by awawa-dev** [reply]
Pico cannot be permanently bricked, the serial port should be available right after Pico automatically restarts after uploading the firmware, I don't use macos and I don't know why you use the screen and not HyperHDR to test it nor if it's Pico id or some other internal serial device.

**Reply by dizzwold** [reply]
@awawa-dev 

 Okay,  on the mac when flashing the board the way described above, there's no way of telling if the flash is complete, no LED indicators or anything.
 I connected the Scorpio to the Pi4 flashed HyperHDR with a usb cable. After a few minuets I see Pico (/dev/ttyAMC0) listed in the Output Path on the Desktop Control Panel.
  In the Logs section I get the following after a few minuets;

2024-07-04T08:50:18.703Z [WEBSOCKET] (HyperAPI.cpp:1109) message: [{"command":"leddevice","ledDeviceType":"adalight","subcommand":"discover","tan":889}]
2024-07-04T08:50:18.759Z [WEBSOCKET] (HyperAPI.cpp:1133) response: [{"devices":[{"name":"Auto","value":"auto"},{"name":"Pico (/dev/ttyACM0)","value":"ttyACM0"}],"ledDeviceType":"adalight"}]

**Reply by dizzwold** [reply]
@awawa-dev ,

  I don't currently have the PSU, so I can't really test with the LED strip, but I understand we have success;
2024-07-04T14:01:32.920Z [LEDDEVICE0_ADALIGHT] DETECTED DEVICE USING HyperSerialEsp8266/HyperSerialESP32/HyperSerialPico FIRMWARE (RGBW => Gain: 255/255, red: 160, green: 160, blue: 160  Welcome! Awa driver 9.) at 3 msec

**Reply by awawa-dev** [reply]
Yes, everything seems correct.

**Comment by dizzwold** [comment]
@awawa-dev 

 Hi Awawa-Dev,


 Just had a reply from the supplier of my HDMI Splitter, which isn't greatly convincing;
_

> LLDV isn't something we've specifically tested for, but it should work fine, as far as I know. The 840BN can't force any modes, as it doesn't participate in the handshake or modify the signal in any way. It tries, as much as possible, to act like a cable between the two devices. A​ny video or audio stream features that fit the bandwidth envelope of the 840BN should work as expected

_
 My PSU arrived today, but I've been out all day so not had chance to wire things up. This will be the only way to find-out if the Kinovo 840BN will push the LLDV through or not.
 I'll try to get things organised to morrow and will report back on the results when I can.

**Comment by awawa-dev** [comment]
Most splitters go the easy way: if DV-capable TV is connected as the main output and a non-DV grabber as the secondary output, standard splitter will report full DV support to the player (because of TV, doesnt care about the grabber). As a result, the grabber will receive a garbage stream because it will not be able to decode it. Ezcoo seems to detect such situations and if the secondary does not support DV it hides DV modes except LLDV in its EDID. At least I think so. They say they haven't tested LLDV, nor is this splitter capable of interfering with EDID handshakes, so be prepared for a negative scenario.

**Comment by dizzwold** [comment]
@awawa-dev 

 Okay. I've got till the end of the month to return it, so we'll give it a go first and report back once tested with the LED Strips.

**Reply by awawa-dev** [reply]
> @awawa-dev
> 
> Okay. I've got till the end of the month to return it, so we'll give it a go first and report back once tested with the LED Strips.

you can test it without the LED strip, only USB grabber is neccesery

**Comment by dizzwold** [comment]
@awawa-dev 

Hi,

 Just tried wiring everything up and I get the following error come up in the log (error 19). Is this the Kinivo Splitter or something else;
`2024-07-09T12:01:53.000Z [SSDP] (SSDPHandler.cpp:42) SSDPHandler is initializing
2024-07-09T12:01:53.001Z [SSDP] (SSDPHandler.cpp:183) The local address is: 192.168.0.174
2024-07-09T12:01:53.001Z [SSDP] (SSDPServer.cpp:95) Initialize the SSDP server
2024-07-09T12:01:53.001Z [WEBSERVER] (StaticFileServing.cpp:68) SSDP description is set up
2024-07-09T12:01:53.001Z [SSDP] Starting the SSDP server
2024-07-09T12:05:55.874Z [WEBSOCKET] (WebSocketClient.cpp:39) New connection from ::ffff:192.168.0.6
2024-07-09T12:05:56.124Z [WEBSOCKET] (HyperAPI.cpp:957) log streaming activated for client ::ffff:192.168.0.6
2024-07-09T12:07:13.909Z [WEBSOCKET] (HyperAPI.cpp:966) log streaming deactivated for client  ::ffff:192.168.0.6
2024-07-09T12:07:25.033Z [WEBSOCKET] (HyperAPI.cpp:1109) message: [{"command":"leddevice","ledDeviceType":"adalight","subcommand":"discover","tan":709}]
2024-07-09T12:07:25.096Z [WEBSOCKET] (HyperAPI.cpp:1133) response: [{"devices":[{"name":"Auto","value":"auto"},{"name":"Pico (/dev/ttyACM0)","value":"ttyACM0"}],"ledDeviceType":"adalight"}]
2024-07-09T12:08:09.856Z [WEBSOCKET] (HyperAPI.cpp:1109) message: [{"command":"leddevice","ledDeviceType":"adalight","subcommand":"discover","tan":151}]
2024-07-09T12:08:09.911Z [WEBSOCKET] (HyperAPI.cpp:1133) response: [{"devices":[{"name":"Auto","value":"auto"},{"name":"Pico (/dev/ttyACM0)","value":"ttyACM0"}],"ledDeviceType":"adalight"}]
2024-07-09T12:13:00.352Z [IMAGETOLED_MNG0] Total index number is: 20520 (memory: 20520). User sparse processing is: disabled, image size: 720 x 480, area number: 108
2024-07-09T12:13:00.353Z [COLORSPACE_CALIBRATION0] classicMode: no, gammas:[1.50, 1.50, 1.50], saturation: 1.00, luminance: 1.00, backLight: [enabled, threshold: 1, colored: yes]
2024-07-09T12:13:00.353Z [LED_CALIBRATION0] (LedCalibration.cpp:101) Calibration config '0' for LED segment: [0, 107]
2024-07-09T12:13:00.353Z [LED_CALIBRATION0] (LedCalibration.cpp:96) LedCalibration is released

> 2024-07-09T12:16:04.024Z [V4L2:AUTO] Throws error nr: VIDIOC_DQBUF error. Video stream is probably broken. Refreshing list of the devices. error code 19, No such device
> 2024-07-09T12:16:04.024Z [V4L2:AUTO] VIDIOC_STREAMOFF  error code  19, No such device

2024-07-09T12:16:04.026Z [V4L2:AUTO] Stopped
2024-07-09T12:16:04.027Z [V4L2:AUTO] Found capture device: /dev/video15
2024-07-09T12:16:04.030Z [V4L2:AUTO] Found capture device: /dev/video21
2024-07-09T12:16:04.033Z [V4L2:AUTO] Found capture device: /dev/video14
2024-07-09T12:16:04.033Z [V4L2:AUTO] Found capture device: /dev/video22
2024-07-09T12:17:56.119Z [WEBSOCKET] (HyperAPI.cpp:957) log streaming activated for client ::ffff:192.168.0.6`

 Some of the LEDs flash On on boot, but I think that's due to the setting I put in Just for testing, so not to worry about that at the moment as I only wanted to test if the Splitter was up to the job.

 Is it the splitter causing the Error 19 VIDIOC_DQBUF?

**Comment by awawa-dev** [comment]
I don't analyze partial logs because it doesn't make sense when a problem like this occurs. Next time, upload it on pastebin.
I suspect that your grabber is behaving quite ugly and unusual and simply disappears from the OS when there is no single video or the source resolution changes (because there is no usually used device `/dev/video0` which was omitted from your logs on beginning). I'm not sure about the role of the HDMI splitter in this case, but it is possible. There is a "revive" option in the grabber option, but with no guarantee it will help because it assumes the grabber will reappear in the OS after 2-3 seconds.

**Comment by dizzwold** [comment]
@awawa-dev 

 Hmm,

 I posted all the log from booting up today. I'd previously booted on the Friday 5/7/24, but with no LED's, No Splitter and No capture card.
 Today was the fist boot with everything connected and copied the log from that point.

 Just booted again and refreshed the browser page. This time the previous boots (from last week), are gone and only today's logs show.

https://pastebin.com/qWNgEq4t

**Comment by awawa-dev** [comment]
So you have what I told you:

```
2024-07-09T12:44:49.673Z [V4L2:AUTO] Found capture device: /dev/video15
2024-07-09T12:44:49.675Z [V4L2:AUTO] Found capture device: /dev/video21
2024-07-09T12:44:49.675Z [V4L2:AUTO] Found capture device: /dev/video14
2024-07-09T12:44:49.676Z [V4L2:AUTO] Found capture device: /dev/video22
2024-07-09T12:44:49.676Z [V4L2:AUTO] Found capture device: /dev/video0
2024-07-09T12:44:49.678Z [V4L2:AUTO] Starting V4L2 grabber. Selected: USB3.0 Video: USB3.0 Video (video0) [/dev/video0] 720 x 480 @ 60 fps yuyv input -1
```
then `/dev/video0` disappears not only from HyperHDR but also from the OS

```
2024-07-09T13:56:21.047Z [V4L2:AUTO] VIDIOC_STREAMOFF  error code  19, No such device
2024-07-09T13:56:21.049Z [V4L2:AUTO] Stopped
2024-07-09T13:56:21.050Z [V4L2:AUTO] Found capture device: /dev/video15
2024-07-09T13:56:21.052Z [V4L2:AUTO] Found capture device: /dev/video21
2024-07-09T13:56:21.053Z [V4L2:AUTO] Found capture device: /dev/video14
2024-07-09T13:56:21.055Z [V4L2:AUTO] Found capture device: /dev/video22
```

**Comment by dizzwold** [comment]
@awawa-dev 

 So is that the splitter or Capture Card or both?

 Something else I've just realised by unplugging HDMI cables and still having the tv playing, I'm not on Full Cable (ie cable tv), I'm on a UHF Aerial so not everything is going through HDMI.

**Comment by awawa-dev** [comment]
> So is that the splitter or Capture Card or both?

the grabber disappears because unfortunately this model behave like that as I told you when there is no video signal or the video signal is changing. try auto-resume option in the grabber.

**Comment by dizzwold** [comment]
@awawa-dev 

 That works!

 So is it worth keeping both the Splitter and Capture card, or should I really go for the Ezcoo and MS2109, MS2130?

 Next to figure my best options for attaching a frame.

**Comment by awawa-dev** [comment]
You have to decide for yourself whether you can live with it despite some drawbacks ;) Ezcoo +ms2130 (+custom firmware) is the most popular choice probably because of Dolby Vision/LLDV, although I don't completely understand the phenomenon of this closed format. It can and does offer quite advanced algorithms, but what if in blind tests people generally prefer HDR in which the colors are more intense and the format itself is open.
